Don't heal damage where there's no ice

einola opened the pull request:
# Don't heal damage where there's no ice

Fixes #895 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

I put an ``if ( cice[i] == 0.) continue;`` statement at the top of ConstantHealing::updateElement to skip ice-free points. This makes sense, but also averts inserting NaNs into the field, as the code does a division by ``cice[i]`` at some point.

---
# Test Description

All tests run.

---
# Documentation Impact

None

---
# Other Details

None

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [ ] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Don't heal damage where there's no ice

I put an ``if ( cice[i] == 0.) continue;`` statement at the top of
ConstantHealing::updateElement to skip ice free points. This makes
sense, but also averts inserting NaNs into the field, as the code does a
division by ``cice[i]`` at some point.

einola committed:
Update new damage when there's no ice

I forgot that there's a new and old damage floating around. So, I should
just copy the old damage value to the new one when there's no ice - it's
not enough to just return.

joewallwork commented on physics/src/modules/DamageHealingModule/ConstantHealing.cpp:
```diff
@@ -51,6 +51,12 @@ void ConstantHealing::update(const TimestepTime& tstep)
 void ConstantHealing::updateElement(size_t i, const TimestepTime& tstep)
 {
+    // No ice, no healing
+    if (cice[i] == 0.) {
```
Might it be worth doing an approximate check? e.g.,
```suggestion
    double tol = 1e-10;
    if (abs(cice[i]) < tol) {
```
Otherwise unexpected behaviour might result when `cice` is effectively (but not quite) zero.

einola replied:
Yes, good point. I'll use IceMinima::c(), which should be used for this purpose throughout the model. In particular, the thermodynamics melts all ice in the grid cell when the concentration drops below this limit.

einola committed:
Don't compare a double against zero

Change the check from cice[i] == 0 to cice[i] <= IceMinima::c(). This
should be more robust, and IceMinima::c() is a reasonable minimum. When
the thermodynamics are running, then all ice melts away when cice[i] <=
IceMinima::c() anyway.

einola committed:
Automatic clang-format fix

Commiting it now, so that it stops nagging me.

joewallwork reviewed (approved):
Great!
